### DISCUSSION TITLE
Is it possible to send shared `Arc<Message>` via the websocket connection?

### DISCUSSION THREAD

**Discussion by sukhmel** [tokio-rs/axum#1976] [Q&A] [answered]
Sorry for bothering again

In my application, I have a lot of broadcasted messages, and a lot of cloning done because of that. I've tried to choose `Arc<Message>` as `SplitSink::Item` so that I have `SplitSink<WebSocket, Arc<Message>` as a sending half, but it doesn't compile

Is there a way not to clone messages that would be broadcast to several WebSocket `SplitSinks`? Maybe I'm just doing it wrong 🤔 

what I'm doing is something like this:
```rust
struct Handler {

    /// Split socket into sender and receiver
    async fn handle_socket(&self, socket: WebSocket) {
        let (sender, _) = socket.split::<Arc<Message>>();
        /* ... */  
        self.senders.insert(key, sender);
    }
    
    /// Try to send the same messages to lots of senders
    async fn broadcast(&self, messages: Vec<String>) {
        let messages = messages
            .into_iter()
            .map(|Message::Text)
            .map(Arc::new)
            .collect::<Vec<_>>();
        for sender in self.senders.values_mut() {
            let mut stream = tokio_stream::iter(messages.clone()).map(anyhow::Ok);
            sender
                .send_all(&mut stream)
                .await
                .unwrap();
        }
    }
}
```

This probably depends on what [tokio-tungstenite](https://github.com/snapview/tokio-tungstenite) is capable of. If so, I will file an issue there also.

Edit: or maybe I need to find another SplitSink implementation 🤔 
compilation errors are as follows:
```
345 | /         sender
346 | |             .send(
    | |             -^^^^ method cannot be called on `SplitSink<WebSocket, Arc<Message>>` due to unsatisfied trait bounds
    | |_____________|


47  |   pub struct SplitSink<S, Item> {
    |   -----------------------------
    |   |
    |   doesn't satisfy `_: Sink<_>`
    |   doesn't satisfy `_: SinkExt<_>`
    |
    = note: the following trait bounds were not satisfied:
            `SplitSink<axum::extract::ws::WebSocket, Arc<axum::extract::ws::Message>>: futures::Sink<_>`
            which is required by `SplitSink<axum::extract::ws::WebSocket, Arc<axum::extract::ws::Message>>: SinkExt<_>`
```

**Comment by davidpdrsn** [comment, marked as answer]
Nope you need to clone the messages.

**Reply by sukhmel** [reply]
As far as I understand from `axum` and `tungstenite` sources, this is because `tungstenite` accepts owned messages, is that correct? Or is there something else that prevents this kind of shared message to be sent?

**Reply by davidpdrsn** [reply]
> As far as I understand from axum and tungstenite sources, this is because tungstenite accepts owned messages, is that correct?

Yes that is correct.

The data needs to be cloned to be written to the socket anyway so I'm not sure using shared messages would help much. If this is something you really need consider asking the tungstenite folk about it.

**Reply by sukhmel** [reply]
yeah, I've found an unmerged PR with a [discussion](https://github.com/snapview/tungstenite-rs/pull/175#issuecomment-1190739028) of problems that arise from sharing data, will leave a link for posterity
